rfd/0078-role-variables.md

```diff
@@ -0,0 +1,439 @@
+other roles will not be visible or effect the current role in any way.
+
+### Can variable definitions depend on other variables?
+
+Yes, within the variable definition block variables may depend on other
+variables only if they are defined *earlier*, or higher-up, in the variable
+list.
+
+It would be technically feasible not to enforce this ordering by sorting the
+variables topologically based on their dependencies, but requiring the ordering
+forces admins to carefully consider their dependencies and automatically avoids
+cycles, which is arguably for the best.
+
+Example:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: example
+spec:
+  vars:
+    - name: var_a
+      values:
+          out: ["foo"]
+    - name: var_b
+      values:
+          out:
+            # var_b can depend on var_a because it is defined above.
+            - "{{vars.var_a}}"
+            # Error: cannot depend on a variable defined later in the list.
+            # - "{{vars.var_c}}"
+    - name: var_c
+      values:
+          # Variables can be used as inputs.
+          input: "{{var_a}}"
+          match: "^foo$"
+          out:
+            - "bar"
+            - "{{vars.var_b}}"
+```
+
+### Trusted clusters
+
+Variables defined in the root cluster will not be usable in leaf clusters.
+It is expected that the variable will be redefined in the leaf cluster.
+
+It's common for deployments to copy the exact same role definitions to all
+clusters, so this seems like the most logical choice.
+As long as the role which defines the variable is present in both clusters and
+it is mapped to its leaf equivalent, you can seamlessly use the variable in both
+clusters.
+
+## Examples
+
+### Defined allow labels for each group
+
+If the Teleport admin cannot create custom traits in their IDP, then it is
+extremely cumbersome and repetitive (borderline impossible) to define a custom
+mapping per group.
+
+Solution today:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users
+spec:
+  allow:
+    node_labels:
+      env:
+        - regexp.replace(external.groups, "dev", "dev")
+        - regexp.replace(external.groups, "dev", "staging")
+        - regexp.replace(external.groups, "qa", "qa")
+        - regexp.replace(external.groups, "qa", "staging")
+    app_labels:
+      env:
+        - regexp.replace(external.groups, "dev", "dev")
+        - regexp.replace(external.groups, "dev", "staging")
+        - regexp.replace(external.groups, "qa", "qa")
+        - regexp.replace(external.groups, "qa", "staging")
+```
+
+With more roles and more complex rules, this would be even more difficult. The
+full list would have to be repeated for every value in every role. I don't think
+anyone even knows you can do this, and I hope no-one actually does, because it's
+ridiculous.
+
+With role variables it becomes trivial:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users
+spec:
+  vars:
+    - name: allow_envs
```

Comment:
So a variable can expand to a single value or a list of values, right?

But there's no difference in syntax, so it doesn't look like we can validate when used incorrectly.

What happens in this example?

- `vars.a` expands to the list containing `["1", "2"]`
- `vars.b` is:
```
out:
- "3"
- "{{vars.a}}"
```

- Is `b` "flattened" to then produce the result `["3", "1", "2"]?
- Would this also be a valid definition for b? `out: {{vars.a}}`?
- If yes, would the same hole true if `a` evaluated to a single value instead of a list?

Reply:
Yes, it will be flattened. A single value is really just a list of length one, it's not really a special case.

This is similar semantically to how traits are used today, lists are always implicitly expanded and flattened. If you have traits `groups = [a,b,c]` then
```yaml
node_labels:
   owner: ["d", "external.groups"]
```
expands to
```yaml
node_labels:
   owner: ["d", "a", "b", "c"]
```

>Would this also be a valid definition for b? `out: {{vars.a}}`
Almost, it would have to be `out: ["{{vars.a}}"]`